### PR TITLE
[#1608][part-2] fix(spark): avoid releasing block in advance when enable block resend

### PR DESCRIPTION
**Pull request by zuston**
### What changes were proposed in this pull request?

1. avoid releasing block previously when enable block resend
2. introduce the block max retry times

### Why are the changes needed?

For: #1608

In the current codebase for partition reassignment, it has some bugs as follows
1. data has been released when resending.
2. if the blocks fail to resend, it may fast fail without retry again

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

`RssShuffleWriterTest#blockFailureResendTest` is to test the resending block mechanism.

### CODE REVIEW

**Comment by zuston** [comment]
PTAL @xumanbu

**Comment by jerqi** [comment]
Could you add more test cases for this pull request? Is this a data correct issue?

**Comment by zuston** [comment]
> Could you add more test cases for this pull request? Is this a data correct issue?

Yes, I'm doing.

 And I'm not satisfied with the current implementation of `one partition -> multi servers` . So the #1608 is to track the bug some fixes and improvements.

And I think we should have the dedicated name to describe this feature, such as `stage retry`. How about `partition data reassign` ?

**Comment by jerqi** [comment]
> 3. fix potential incorrect block checking logic


Could you elabrate more about point 3?

**Comment by zuston** [comment]
> > 3. fix potential incorrect block checking logic
> 
> Could you elabrate more about point 3?

My fault, the previous checking logic is right

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java [hunk 1]
```diff
@@ -427,20 +425,16 @@ public List<AddBlockEvent> buildBlockEvents(List<ShuffleBlockInfo> shuffleBlockI
...
-        final List<ShuffleBlockInfo> shuffleBlocksTemp = shuffleBlockInfosPerEvent;
         events.add(
             new AddBlockEvent(
                 taskId,
                 shuffleBlockInfosPerEvent,
-                () -> {
-                  freeAllocatedMemory(memoryUsedTemp);
-                  shuffleBlocksTemp.stream().forEach(x -> x.getData().release());
+                (block, isSuccessful) -> {
+                  this.freeAllocatedMemory(block.getFreeMemory());
```
Comment (xumanbu) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:433]: free memory logic may check if `isSuccessful` is true?

File: common/src/main/java/org/apache/uniffle/common/exception/RssSendFailedException.java [hunk 2]
```diff
@@ -18,6 +18,11 @@
 package org.apache.uniffle.common.exception;
 
 public class RssSendFailedException extends RssException {
+
+  public RssSendFailedException() {
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/exception/RssSendFailedException.java:22]: Could we use `super()`? This is not a common practice.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java (hunk 1 above)
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:433]: Needn't.

**Review by xumanbu — APPROVED** [review]

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/AddBlockEvent.java [hunk 3]
```diff
@@ -59,6 +59,23 @@ public List<Runnable> getProcessedCallbackChain() {
...
+  public void withBlockProcessedCallback(
+      TupleConsumer<ShuffleBlockInfo, Boolean> blockProcessedCallback) {
+    this.blockProcessedCallback = blockProcessedCallback;
+  }
+
+  public TupleConsumer<ShuffleBlockInfo, Boolean> getBlockProcessedCallback() {
+    return blockProcessedCallback;
+  }
+
+  public void enableBlockResend() {
```
[the remark this answers is not in the document]
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/AddBlockEvent.java]: Removed. 

Done.

File: common/src/main/java/org/apache/uniffle/common/ShuffleBlockInfo.java [hunk 4]
```diff
@@ -84,6 +86,14 @@ public ShuffleBlockInfo(
     this.taskAttemptId = taskAttemptId;
   }
 
+  public int getRetryCounter() {
+    return retryCounter;
+  }
+
+  public void setRetryCounter(int retryCounter) {
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/ShuffleBlockInfo.java:93]: Do we need this method? `setCounter` seems weird. `increateCounter` or `resetCounter`?

Comment (zuston) [reply common/src/main/java/org/apache/uniffle/common/ShuffleBlockInfo.java:93]: Please see the invoking code.

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java [hunk 5]
```diff
@@ -418,89 +433,108 @@ protected void checkBlockSendResult(Set<Long> blockIds) {
...
+    if (failedBlockIds == null || failedBlockIds.isEmpty()) {
+      return;
     }
-    return resendBlockStatusSet;
+
+    Set<TrackingBlockStatus> resendCandidates = new HashSet<>();
+    // to check whether the blocks resent exceed the max resend count.
+    for (Long blockId : failedBlockIds) {
+      List<TrackingBlockStatus> retryRecords = failedTracker.getFailedBlockStatus(blockId);
+      // todo: support retry times by config
```
Comment (jerqi) [inline client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:474]: Could you create an issue to track this?

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java [hunk 6]
```diff
@@ -418,89 +433,108 @@ protected void checkBlockSendResult(Set<Long> blockIds) {
...
-    Set<Long> failedBlockIds = shuffleManager.getFailedBlockIds(taskId);
-    if (taskFailRetry && !failedBlockIds.isEmpty()) {
-      Set<TrackingBlockStatus> shouldResendBlockSet = shouldResendBlockStatusSet(failedBlockIds);
-      try {
-        reSendFailedBlockIds(shouldResendBlockSet);
-      } catch (Exception e) {
-        LOG.error("resend failed blocks failed.", e);
+  private void dataCheckOrRetry() {
+    if (isBlockFailSentRetryEnabled) {
+      collectBlocksToResendOrFastFail();
```
Comment (jerqi) [inline client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:438]: Em... Could you simplify the logic?

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java (hunk 5 above)
Comment (zuston) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:474]: It will be created after all done

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java [hunk 7]
```diff
@@ -419,89 +434,108 @@ protected void checkBlockSendResult(Set<Long> blockIds) {
...
-                            failedBlockInfo.getData(),
-                            Lists.newArrayList(finalDynamicShuffleServer),
-                            failedBlockInfo.getUncompressLength(),
-                            failedBlockInfo.getFreeMemory(),
-                            taskAttemptId));
-                  });
+              for (TrackingBlockStatus blockStatus : failedBlockStatusSet) {
+                ShuffleBlockInfo failedBlockInfo = blockStatus.getShuffleBlockInfo();
+                failedBlockInfoList.add(failedBlockInfo);
+                ShuffleBlockInfo newBlock =
```
Comment (jerqi) [inline client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:518]: Why do we use a new block instead of reusing old block?

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/DataPusher.java [hunk 8]
```diff
@@ -88,14 +88,25 @@ public CompletableFuture<Long> send(AddBlockEvent event) {
...
-            SendShuffleDataResult result =
+            result =
                 shuffleWriteClient.sendShuffleData(
                     rssAppId, shuffleBlockInfoList, () -> !isValidTask(taskId));
             putBlockId(taskToSuccessBlockIds, taskId, result.getSuccessBlockIds());
             putFailedBlockSendTracker(
                 taskToFailedBlockSendTracker, taskId, result.getFailedBlockSendTracker());
           } finally {
+            Set<Long> succeedBlockIds = result.getSuccessBlockIds();
+            if (succeedBlockIds != null && !succeedBlockIds.isEmpty()) {
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/writer/DataPusher.java:101]: CollectionUtils has a similar method to simplify code.

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java [hunk 9]
```diff
@@ -419,89 +434,108 @@ protected void checkBlockSendResult(Set<Long> blockIds) {
...
+      return;
     }
-    return resendBlockStatusSet;
+
+    Set<TrackingBlockStatus> resendCandidates = new HashSet<>();
+    // to check whether the blocks resent exceed the max resend count.
+    for (Long blockId : failedBlockIds) {
+      List<TrackingBlockStatus> retryRecords = failedTracker.getFailedBlockStatus(blockId);
+      // todo: support retry times by config
+      if (retryRecords.size() >= blockFailSentMaxTimes) {
```
Comment (jerqi) [inline client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:476]: Do we need retryCounter if we have retryRecords?

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java (hunk 7 above)
Comment (zuston) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:518]: Sounds good. Let me optimize this tomorrow

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/DataPusher.java (hunk 8 above)
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/DataPusher.java:101]: Got it

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java (hunk 9 above)
Comment (zuston) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:476]: Let me think again tomorrow

**Comment by jerqi** [comment]
Could you ensure the logic don't involve memory leak?

**Comment by jerqi** [comment]
@rickyma Could you help review this pull request?

**Comment by zuston** [comment]
> Could you ensure the logic don't involve memory leak?

Just using the unit tests to ensure this.

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java (hunk 9 above)
Comment (zuston) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:496]: Done

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java [hunk 10]
```diff
@@ -419,89 +434,108 @@ protected void checkBlockSendResult(Set<Long> blockIds) {
...
-      try {
-        reSendFailedBlockIds(shouldResendBlockSet);
-      } catch (Exception e) {
-        LOG.error("resend failed blocks failed.", e);
+  private void checkDataIfAnyFailure() {
+    if (isBlockFailSentRetryEnabled) {
+      collectFailedBlocksToResend();
+    } else {
+      if (hasAnyBlockFailure()) {
+        throw new RssSendFailedException("Send fail");
```
[the remark this answers is not in the document]
Comment (zuston) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java]: Done

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/DataPusher.java (hunk 8 above)
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/DataPusher.java]: Done

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java (hunk 7 above)
Comment (zuston) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:538]: Emm, I reserve this logic becuase the old block will be used in the other places.

File: common/src/main/java/org/apache/uniffle/common/exception/RssSendFailedException.java [hunk 11]
```diff
@@ -18,6 +18,7 @@
 package org.apache.uniffle.common.exception;
 
 public class RssSendFailedException extends RssException {
+
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/exception/RssSendFailedException.java:21]: An unnecessary change?

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java (hunk 7 above)
Comment (jerqi) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:538]: It probably involve the issues if we use a new block. It's not a good practice. When I saw the this pattern, I will worry about the release of data and management of data. Maybe you use other code to guarantee this, but when seeing similar code, I will worry about the issues which I mentioned. So if it's possible, I prefer not using this pattern.

Comment (zuston) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:538]: Make sense.

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java [hunk 12]
```diff
@@ -113,7 +114,8 @@ public class RssShuffleWriter<K, V, C> extends ShuffleWriter<K, V> {
   private final Set<Long> blockIds = Sets.newConcurrentHashSet();
   private TaskContext taskContext;
   private SparkConf sparkConf;
-  private boolean blockSendFailureRetryEnabled;
+  private boolean isBlockFailSentRetryEnabled;
```
Comment (jerqi) [inline client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:117]: `isBlockFailSentRetryEnabled ` -> `blockFailSentRetryEnabled `

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java [hunk 13]
```diff
@@ -363,6 +367,34 @@ protected List<CompletableFuture<Long>> postBlockEvent(
...
+                bufferManager.freeAllocatedMemory(block.getFreeMemory());
+                block.getData().release();
+                return;
+              }
+
+              // for the exceeding max send retry blocks, we should release its data and allocated
+              // memory.
+              List<TrackingBlockStatus> record = tracker.getFailedBlockStatus(block.getBlockId());
+              if (CollectionUtils.isEmpty(record)) {
+                LOG.error(
```
Comment (jerqi) [inline client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:384]: Do we need to add an assert here?

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java [hunk 14]
```diff
@@ -422,45 +454,63 @@ protected void checkBlockSendResult(Set<Long> blockIds) {
...
-          .forEach(trackingBlockStatus -> resendBlockStatusSet.add(trackingBlockStatus));
+  private void collectFailedBlocksToResend() {
+    if (!isBlockFailSentRetryEnabled) {
+      return;
     }
-    return resendBlockStatusSet;
+
+    FailedBlockSendTracker failedTracker = shuffleManager.getBlockIdsFailedSendTracker(taskId);
+    Set<Long> failedBlockIds = failedTracker.getFailedBlockIds();
+    if (failedBlockIds == null || failedBlockIds.isEmpty()) {
```
Comment (jerqi) [inline client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:487]: Could you use CollectionUtils here, too?

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java [hunk 15]
```diff
@@ -694,4 +745,14 @@ private void throwFetchFailedIfNecessary(Exception e) {
...
   }
+
+  @VisibleForTesting
+  protected void enableBlockFailSentRetry() {
+    this.isBlockFailSentRetryEnabled = true;
+  }
+
+  @VisibleForTesting
+  protected void addReassignmentShuffleServer(String shuffleId, ShuffleServerInfo replacement) {
+    failoverShuffleServers.put(shuffleId, replacement);
```
Comment (jerqi) [inline client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:756]: `faultServers` or `replacementServer`? What's the meaning of `failoverServers`?

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/AddBlockEvent.java [hunk 16]
```diff
@@ -59,6 +61,15 @@ public List<Runnable> getProcessedCallbackChain() {
     return processedCallbackChain;
   }
 
+  public void withBlockProcessedCallback(
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/writer/AddBlockEvent.java:64]: `BlockSentCallback` may be better?

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java (hunk 15 above)
Comment (zuston) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java]: done

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java (hunk 14 above)
Comment (zuston) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java]: done

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java (hunk 13 above)
Comment (zuston) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java]: done

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java (hunk 12 above)
Comment (zuston) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java]: done

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java (hunk 5 above)
Comment (zuston) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:483]: After rethinking ,this is wrong of using retryRecords. I have corrected this part.

**Comment by zuston** [comment]
All done @jerqi PTAL

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/AddBlockEvent.java [hunk 17]
```diff
@@ -21,12 +21,14 @@
...
 
 import org.apache.uniffle.common.ShuffleBlockInfo;
+import org.apache.uniffle.common.function.TupleConsumer;
 
 public class AddBlockEvent {
 
   private String taskId;
   private List<ShuffleBlockInfo> shuffleDataInfoList;
   private List<Runnable> processedCallbackChain;
+  private TupleConsumer<ShuffleBlockInfo, Boolean> blockSentCallback;
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/writer/AddBlockEvent.java:31]: Could we give a more unify design between processedCallbackChain and blockSentCallback?

Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/AddBlockEvent.java:31]: I will remove processedCallbackChain in the future, it looks unnecessary.

Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/AddBlockEvent.java:31]: From current usage, needn't

Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/AddBlockEvent.java:31]:
```
interface BlockFailureCallback {
onBlockFailure(ShuffleBlock block);
}

interface BlockSuccessCallback {
onBlockSuccess(ShuffleBlock block);
}

interface BlockCompletionCallback {
onBlockCompletion(ShuffleBlock block);
}
```

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/AddBlockEvent.java [hunk 18]
```diff
@@ -61,16 +54,30 @@ public List<Runnable> getProcessedCallbackChain() {
...
+    if (block == null) {
+      return;
+    }
+    this.blockFailureCallback.onBlockFailure(block);
   }
 
-  @Override
-  public String toString() {
-    return "AddBlockEvent: TaskId[" + taskId + "], " + shuffleDataInfoList;
+  public void callbackOnBlockSuccess(ShuffleBlockInfo block) {
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/writer/AddBlockEvent.java:69]: The name of the method seems weird. Let us think a better name.

Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/AddBlockEvent.java]: done

**Comment by jerqi** [comment]
After reading the release process, I think we have multiple ways to release the memory. FailureCallback, fast fail and so on. Could you have a common release process for them. Maybe we can release in the method `close`.
Another point is whether SuccessCallback is suitable field in the Event. WDYT?

**Comment by zuston** [comment]
> After reading the release process, I think we have multiple ways to release the memory. FailureCallback, fast fail and so on. Could you have a common release process for them. Maybe we can release in the method close.

For this case, the way of memory release on failure and success is mutually exclusive. That means the close method is not general and abstract to express this. And the close for the event is really weird, although I understand your thought. But this thought may be sensible on the having ownership mechanism of Rust

> Another point is whether SuccessCallback is suitable field in the Event. WDYT?

It's blockSuccessCallback, rather than eventSuccessCallback. So it's OK for me.

**Comment by jerqi** [comment]
A native code:
```
try {
   sendBlocks();
} catch (Exception e) {

} finally {
    if (block : successBlocks) {
        block.triggerSuccessfulCallback(); // triggerSuccessfulCallback to release data
    }
    if (block: failureBlocks) {
        block.triggerFailureCallback(); // trggerFailureCallback to resend data if the retry times don't reach the max value.
    }
}
```

```
BufferManager {

void close() {
     for (block : flightingBlocks) {
         block.triggerSuccessCallback();
     }
}
}
```

**Comment by rickyma** [comment]
> @rickyma Could you help review this pull request?

Currently, the Netty part of the code is running stably, and we have gone through multiple rounds of testing, with the number of issues narrowed down to single digits. Regarding the PR related to https://github.com/apache/incubator-uniffle/issues/1608, we only hope that **_it does not affect the original code logic and cause a decrease in stability_**. Otherwise, we will still need to spend extra time re-testing, which is not very likely at this stage, since we're planning to deploy to production in mid-April.

@jerqi @zuston

**Comment by zuston** [comment]
> > @rickyma Could you help review this pull request?
> 
> Currently, the Netty part of the code is running stably, and we have gone through multiple rounds of testing, with the number of issues narrowed down to single digits. Regarding the PRs related to #1608, we only hope that **_they do not affect the original code logic and cause a decrease in stability_**. Otherwise, we will still need to spend extra time re-testing, which is not very likely at this stage, since we're planning to deploy to production in mid-April.
> 
> @jerqi @zuston

I understand your thought, but I have to say that this is an open source project although I hope all changes could not effect the performance and compatiblitiy, it's hard to do.

Based upon the reasons, If you want to meet your internal requirements or stability, you have to take care the commits change and maintain the internal stable version or review code actively.

**Comment by rickyma** [comment]
> I understand your thought, but I have to say that this is an open source project although I hope all changes could not effect the performance and compatiblitiy, it's hard to do.

This issue https://github.com/apache/incubator-uniffle/issues/1608 seems to involve quite a few changes.
If possible, I hope we can provide more detailed descriptions of the proposed changes in the issue and related PRs, explain why we are making these changes **_in detail_**, and provide a clear background of the problem **_in detail_** for future reference and to facilitate a clearer code review for everyone involved.
The current issue and related PR descriptions seem to be a bit too simplistic... especially as time goes by, it becomes inconvenient to understand the background.

**Comment by rickyma** [comment]
I think the following issues and PRs provide more detailed descriptions of the background and solutions, which can be used as good reference examples:

PR: https://github.com/apache/incubator-uniffle/pull/1565, https://github.com/apache/incubator-uniffle/pull/1574
issue: https://github.com/apache/incubator-uniffle/issues/1621, https://github.com/apache/incubator-uniffle/issues/1433

**Comment by zuston** [comment]
> > I understand your thought, but I have to say that this is an open source project although I hope all changes could not effect the performance and compatiblitiy, it's hard to do.
> 
> This issue #1608 seems to involve quite a few changes. If possible, I hope we can provide more detailed descriptions of the proposed changes in the issue and related PRs, explain why we are making these changes **_in detail_**, and provide a clear background of the problem **_in detail_** for future reference and to facilitate a clearer code review for everyone involved. The current issue and related PR descriptions seem to be a bit too simplistic... especially as time goes by, it becomes inconvenient to understand the background.

This beta feature is still on-going(not a producation feature) and if you want to dig more, you can find more from the issue link. But if you want to know why this happen, I think the test case will show somthing you want to know.

And for your reviewing if you want, feel free to discuss more about the design and fix / improvement.

**Comment by zuston** [comment]
> A native code:
> 
> ```
> try {
>    sendBlocks();
> } catch (Exception e) {
> 
> } finally {
>     if (block : successBlocks) {
>         block.triggerCompleteCallback(); // triggerCompleteCallback to release data
>     }
>     if (block: failureBlocks) {
>         block.triggerFailureCallback(); // trggerFailureCallback to resend data if the retry times don't reach the max value.  if the failure reaches the max value, we trigger complete callback. Maybe we should also put fast logic into this function.
>     }
> }
> ```
> 
> ```
> BufferManager {
> 
> void close() {
>      for (block : flightingBlocks) {
>          block.triggerCompleteCallback();
>      }
> }
> ```

I have introduced the block level completion callback.

File: client-spark/spark3/src/test/java/org/apache/spark/shuffle/writer/RssShuffleWriterTest.java [hunk 19]
```diff
@@ -161,8 +355,8 @@ public void checkBlockSendResultTest() {
         assertThrows(
             RuntimeException.class,
             () -> rssShuffleWriter.checkBlockSendResult(Sets.newHashSet(1L, 2L, 3L)));
-    System.out.println(e2.getMessage());
-    assertTrue(e3.getMessage().startsWith("Send failed:"));
+    System.out.println(e3.getMessage());
```
Comment (jerqi) [inline client-spark/spark3/src/test/java/org/apache/spark/shuffle/writer/RssShuffleWriterTest.java:358]: Remove `System.out.prinltln`?

Comment (zuston) [reply client-spark/spark3/src/test/java/org/apache/spark/shuffle/writer/RssShuffleWriterTest.java]: Done.

**Review by jerqi — APPROVED** [review]
LGTM, thanks @zuston .

**Comment by zuston** [comment]
Thanks @jerqi @xumanbu @rickyma  for your review and thoughts. Merged.
